airflow/providers/grafana/log/loki_task_handler.py

```diff
@@ -0,0 +1,242 @@
+
+    def convert_epoch_time(self, timestamp: str) -> int:
+        """Convert timestamp to epoch"""
+        return int(time.mktime(time.strptime(timestamp[:19], self.timestamp_pattern)))
+
+    def query_time_range(
+        self,
+        task_instance: TaskInstance,
+    ) -> Tuple[int, int]:
+        """
+        Return task start and end time in epoch format for query.
+        If Task's try_number > 1 then it will use task execution time as start time.
+        :param ti: task instance object
+        :type: task_instance: TaskInstance
+        """
+        start_time, end_time, execution_time = (
+            str(task_instance.start_date.isoformat()),
+            task_instance.end_date,
+            str(task_instance.execution_date.isoformat()),
+        )
+        next_try = task_instance.next_try_number
+        start_epoch_time = self.convert_epoch_time(start_time)
+
+        if end_time and next_try < 3:
+            end_epoch_time = self.convert_epoch_time(end_time.isoformat())
+            return (start_epoch_time - 60, end_epoch_time + 60 * 2)
+        if end_time:
+            end_epoch_time = self.convert_epoch_time(end_time.isoformat())
+            execution_time_epoch_time = self.convert_epoch_time(execution_time)
+            return (execution_time_epoch_time - 60, end_epoch_time + 60 * 2)
+        return (start_epoch_time - 60, start_epoch_time + 60 * 60 * 24)
+
+    def get_read_parameters(self, task_instance, try_number) -> Dict[str, Union[str, int]]:
+        """
+        Construct query(LogQL) and update parameters for loki read request.
+        :param ti: task instance object
+        :type: task_instance: TaskInstance
+        :param try_number: current try_number to read log from
+        :type: try_number: int
+        """
+        start_time, end_time = self.query_time_range(task_instance)
+        task_lables = self.get_label(task_instance)
+        del task_lables["airflow_try_number"]  # try_number will come from read method
+
+        # converting task_lables(dict) to LogQL format
+        # {"dag_id":"airflow"} --> {dag_id = "airflow"}
+        query = "{"
+        for key, value in task_lables.items():
+            query += f"""{key}="{value}", """
+        query += f"""airflow_try_number="{try_number}"}} != "WARNING" """
+
+        return {
+            "direction": "forward",
+            "start": start_time,
+            "end": end_time,
+            "query": query,
+            "limit": 5000,
+        }
+
+    def _read(self, task_instance, try_number, metadata=None) -> Tuple[str, Dict[str, bool]]:
+        """
+        Return list of logs string for each task try number by streaming it from Loki
+        :param ti: task instance object
+        :type: task_instance: TaskInstance
+        :param try_number: current try_number to read log from
+        :type: try_number: int
+        """
+        logs = ""
+        if self.is_loki_alive() == 200:
```

Comment:
IMO this type of pre-is-alive-check is redundant, better just try the real query right away and only fallback on any exceptions.

Checking for is_alive before:
* Introduces unnecessary delay
* Introduces potential race condition where alive at first works, but 1ms later server goes down and query fails. Forcing you to deal with the query-failure scenario anyway.
* Introduces one more source of error, in case the is_alive fails because of intermittent network flakyness.